# Recombinate User guide:
This set of scripts analyzes sequencing reads to identify or visualize recombination events between two or more viruses. Reads are 'windowed', each window is aligned to the given references, and a relative score is given for each reference and window. 

- 1) Generate a single reference fasta file containing all the references you would like to check for recombinations. For best results, the reads/sequences should primarily be reads you suspect may be recombinations.
- 2) Edit the `recombinate.sh` script to your specifications:
   - A) db_file: Path to the reference file
   - B) read_file: Path to the sequencing reads
   - C) window_length: Size of window in bases
   - D) window_overlap: Number of bases to overlap each window (0-length of window; default=0)
   - E) prefix: Prefix to add to final dataset
   - F) blast_task: blast n task (blastn' 'blastn-short' 'dc-megablast' 'megablast' 'rmblastn')
- 3) Run script with `bash recombinate.sh`
- 4) After bash script is completed, run `recombo_graph()` function in R:
   - A) path="recombo" : Path to read directories generated by `recombinate.sh`
   - B) save_it = T, : Option to save the dataset or not
   - C) graph_it = T : Option to graph to outputs or not
   - D) save_widget = T : Option to save the final summary graph as a plotly widget
   - E) prefix = "" : Prefix to add to final datasets
